core/src/toga/command.py

```diff
@@ -330,6 +366,23 @@ def clear(self):
         if self.on_change:
             self.on_change()
 
+    @contextmanager
+    def suspend_updates(self):
+        """Temporarily suspend ``on_change`` updates.
+
+        Inside this context manager, the ``on_change`` handler will not be
+        invoked if the CommandSet is modified. The ``on_change`` handler (if it
+        exists0 will be invoked when the context manager exits.
+        """
+        orig_on_change = self.on_change
+        self.on_change = None
+        try:
+            yield
+        finally:
+            self.on_change = orig_on_change
+            if self.on_change:
+                self.on_change()
+
```

Comment:
Out of curiosity, what is the intended use-case or concern that this is addressing? Do we have something similar for other classes with `on_change` handlers?

Reply:
It's an optimisation for app startup. On Cocoa, the app instance is created and configured as part of the App constructor. The last thing the app constructor does is install an on_change handler on the app's command set, so on_change is only triggered when there is an actual change after initial setup.

However, on GTK, the app is *created* during the constructor, but isn't *configured* until the GTK `startup` signal is received. This means the on_change handler is installed *before* the app commands have been fully constructed, so you end up invoking create_menus multiple times as the user's startup menu is invoked. 

An analogous problem exists on Winforms; the user's `startup()` method isn't invoked until the main loop is started.

Thinking about it with a clear head, I guess an alternate approach would be to introduce a "finalise app construction" method on the core app, so that anything that needs to happen once the user's configuration code has all executed, and split core's constructor into two, using the call to the factory that constructs the App impl as the split point. At present, the only thing that needs to be done "late" is installing the on_change handler, but I guess there could be others.

Reply:
Thanks for the explanation!

I think my preference would be to indeed keep this only as a public API if you think it is useful to toga users, for example when modifying commands after startup, and not because of implementation details for some backends.

Reply:
That's reasonable. I think there's definitely a use case for a public "suspend" API for events as an optimisation, but that needs a more general approach. Plus, reworking startup() so there's a "finalise" stage makes more architectural to me sense anyway.